File: Exámenes/Finales/FEB 2016/haskell/DataStructures/Graph/EulerianCycle.hs
-------------------------------------------------------------------------------
-- Student's name:
-- Student's group:
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module DataStructures.Graph.EulerianCycle(isEulerian, eulerianCycle) where

import DataStructures.Graph.Graph
import Data.List

--H.1)
isEulerian :: Eq a => Graph a -> Bool
isEulerian g 
    | isEmpty g = True
    | length (vertices g) == 1 = True
    | otherwise = aux vertices g
    where 
        aux [] = True
        aux (v:vs) = even (degree v g) && aux vs

-- H.2)
remove :: (Eq a) => Graph a -> (a,a) -> Graph a
remove g (v,u) = aux (deleteEdge g (v,u))
    where
        aux g
            | (degree v == 0) && (degree u == 0) = delete v (delete u g)
            | (degree v == 0) = delete v g
            | (degree u == 0) = delete u g

-- H.3)
extractCycle :: (Eq a) => Graph a -> a -> (Graph a, Path a)
extractCycle g v0 = undefined

-- H.4)
connectCycles :: (Eq a) => Path a -> Path a -> Path a
connectCycles xs (y:ys)  = undefined

-- H.5)
vertexInCommon :: Eq a => Graph a -> Path a -> a
vertexInCommon g cycle = undefined

-- H.6) 
eulerianCycle :: Eq a => Graph a -> Path a
eulerianCycle g = undefined

File: Exámenes/2016 FEB/prEurlerianCycles2/src/dataStructures/graph/EulerianCycle.java
/**
 * Student's name: Ángel Manuel Soria Gil
 * Student's group: Doble Grado Matemáticas e Ingeniería Informática
 * Data Structures. Grado en Informática. UMA.
 */

package dataStructures.graph;

import dataStructures.list.*;

import java.util.Iterator;

public class EulerianCycle<V> {
    private List<V> eCycle;

    @SuppressWarnings("unchecked")
    public EulerianCycle(Graph<V> g) {
        Graph<V> graph = (Graph<V>) g.clone();
        eCycle = eulerianCycle(graph);
    }

    public boolean isEulerian() {
        return eCycle != null;
    }

    public List<V> eulerianCycle() {
        return eCycle;
    }

    // J.1
    private static <V> boolean isEulerian(Graph<V> g) {
        boolean res = true;
        for (V v : g.vertices()){
            res = res && g.degree(v)%2==0;
        }
        return res;
    }

    // J.2
    private static <V> void remove(Graph<V> g, V v, V u) {
        g.deleteEdge(v, u);
        if (g.degree(v)<1){
            g.deleteVertex(v);
        }
        if (g.degree(u)<1){
            g.deleteVertex(u);
        }
    }

    // J.3
    private static <V> List<V> extractCycle(Graph<V> g, V v0) {
        List<V> res = new LinkedList<>();
        Iterator<V> it = g.successors(v0).iterator();
        V v = v0;
        V u = it.next();
        res.append(u);
        remove(g, v, u);
        while (!u.equals(v0)){
            it = g.successors(u).iterator();
            v = u;
            u = it.next();
            res.append(u);
            remove(g, v, u);
        }
        return res;
    }

    // J.4
    public static <V> void connectCycles(List<V> xs, List<V> ys) {
        if(xs.isEmpty()){
            xs = ys;
        }else{
            List<V> lista = new ArrayList<>();
            Iterator<V> it = xs.iterator();
            while(it.hasNext()){
                V v = it.next();
                lista.append(v);
                if(v==ys.get(0)){
                    for(V y : ys){
                        lista.append(y);
                    }
                }
            }
            xs = lista;
        }
    }

    // J.5
    private static <V> V vertexInCommon(Graph<V> g, List<V> cycle) {
        for (V v : cycle){
            if (g.vertices().isElem(v)){
                return v;
            }
        }
        return null;
    }

    // J.6
    private static <V> List<V> eulerianCycle(Graph<V> g) {
        if (!isEulerian(g)) {
            return null;
        }
        if (g.isEmpty()){
            return new LinkedList<>();
        }
        List<V> res;
        Iterator<V> it = g.vertices().iterator();
        V v = it.next();
        res = extractCycle(g, v);
        while (!g.isEmpty()){
            V v1 = vertexInCommon(g, res);
            List<V> cycle = extractCycle(g, v1);
            connectCycles(res, cycle);
        }
        return res;
    }
}
